src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestStream.cs

```diff
@@ -148,38 +177,55 @@ private void CheckDisposed()
             }
         }
 
-        private void WriteInternal(byte[] buffer, int offset, int count)
+        private async Task InternalWriteAsync(byte[] buffer, int offset, int count, CancellationToken token)
         {
             if (_chunkedMode)
             {
                 string chunkSizeString = String.Format("{0:x}\r\n", count);
                 byte[] chunkSize = Encoding.UTF8.GetBytes(chunkSizeString);
 
-                WriteData(chunkSize, 0, chunkSize.Length);
+                await InternalWriteDataAsync(chunkSize, 0, chunkSize.Length, token);
 
-                WriteData(buffer, offset, count);
-                WriteData(s_crLfTerminator, 0, s_crLfTerminator.Length);
+                await InternalWriteDataAsync(buffer, offset, count, token);
+                await InternalWriteDataAsync(s_crLfTerminator, 0, s_crLfTerminator.Length, token);
             }
             else
             {
-                WriteData(buffer, offset, count);
+                await InternalWriteDataAsync(buffer, offset, count, token);
             }
         }
 
-        private void WriteData(byte[] buffer, int offset, int count)
+        private Task<bool> InternalWriteDataAsync(byte[] buffer, int offset, int count, CancellationToken token)
         {
-            uint bytesWritten = 0;
-            GCHandle pinnedHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned);
-            bool result = Interop.WinHttp.WinHttpWriteData(
-                _requestHandle,
-                Marshal.UnsafeAddrOfPinnedArrayElement(buffer, offset),
-                (uint)count,
-                out bytesWritten);
-            pinnedHandle.Free();
-            if (!result)
+            // TODO (Issue 2505): replace with PinnableBufferCache.
+            if (!_cachedSendPinnedBuffer.IsAllocated || _cachedSendPinnedBuffer.Target != buffer)
             {
-                throw new IOException(SR.net_http_io_read, WinHttpException.CreateExceptionUsingLastError());
+                if (_cachedSendPinnedBuffer.IsAllocated)
+                {
+                    _cachedSendPinnedBuffer.Free();
+                }
```

Comment:
Is it possible for two operations to be issued concurrently?  If not, this looks fine.  If it is, then we're potentially free'ing the pinning handle on a buffer that may still be in use.  I ask just because I'm not sure in what situation you may end up with a different buffer than what's already pinned.